### PR TITLE
Support cases like "within / less than / more than <duration date>"

### PR DESCRIPTION
**Pull request by guom08**
#576

### CODE REVIEW

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateExtractor.cs [hunk 1]
```diff
@@ -306,9 +306,16 @@ private List<Token> DurationWithBeforeAndAfter(string text, DateObject reference
             foreach (var er in durationEr)
             {
                 // if it is a multiple duration and its type is not equal to Date than skip it.
-                if (er.Data != null && er.Data.ToString() != Constants.MultipleDuration_Date)
+                if (er.Data != null && er.Data.ToString().StartsWith("multipleDuration") && er.Data.ToString() != Constants.MultipleDuration_Date)
```
Comment (guom08) [inline .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateExtractor.cs]: Cases like "more than 3 days from today" will result in er.Data equals to "more".
The original condition can't handle this case, so add more accurate restrictions here.

File: .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs [hunk 2]
```diff
@@ -160,8 +160,8 @@ public static class DateTimeDefinitions
 		public const string PeriodTimeOfDayRegex = @"\b((in\s+(the)?\s+)?((?<early>early(\s+|-))|(?<late>late(\s+|-)))?(?<timeOfDay>morning|afternoon|night|evening))\b";
 		public static readonly string PeriodSpecificTimeOfDayRegex = $@"\b(({StrictRelativeRegex}\s+{PeriodTimeOfDayRegex})\b|\btonight)\b";
 		public static readonly string PeriodTimeOfDayWithDateRegex = $@"\b(({TimeOfDayRegex}(\s+(on|of))?))\b";
-		public const string LessThanRegex = @"\b(less\s+than)\s+$";
-		public const string MoreThanRegex = @"\b(more\s+than)\s+$";
+		public const string LessThanRegex = @"\b(less\s+than)\b";
```
Comment (guom08) [inline .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs]: These two Regexes now not only apply to cases like "3 days or more than"
But also "more than 3 days"

File: .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs [hunk 3]
```diff
@@ -193,7 +193,7 @@ public static class DateTimeDefinitions
 		public const string AgoRegex = @"\b(ago|before (?<day>yesterday|today))\b";
 		public const string LaterRegex = @"\b(later|from now|(from|after) (?<day>tomorrow|tmr|today))\b";
 		public const string InConnectorRegex = @"\b(in)\b";
-		public static readonly string WithinNextPrefixRegex = $@"\b(within(\s+the)?(\s+{NextPrefixRegex})?)\b";
+		public static readonly string WithinNextPrefixRegex = $@"\b(within(\s+the)?(\s+(?<next>{NextPrefixRegex}))?)\b";
```
Comment (guom08) [inline .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs]: This change is used to filter out cases like "within the next 5 days before today"

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateExtractor.cs [hunk 4]
```diff
@@ -306,9 +306,16 @@ private List<Token> DurationWithBeforeAndAfter(string text, DateObject reference
...
                 {
                     continue;
+                }
+
+                // special treatment for cases like "more than 3 days from today"
```
Comment (guom08) [inline .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateExtractor.cs]: For cases like "more than 3 days from today", this method should extract "3 days from today" (duration).
But the DurationExtractor will extract "more than 3 days" first.
Here we add special treatment for this "more than 3 days" to let this method work as expected.
This is not quite beautiful here, but seems no good way.

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDatePeriodExtractor.cs [hunk 5]
```diff
@@ -189,6 +190,32 @@ private List<Token> SingleTimePointWithPatterns(string text, DateObject referenc
                     string beforeString = text.Substring(0, (int)extractionResult.Start);
                     ret.AddRange(GetTokenForRegexMatching(beforeString, config.WeekOfRegex, extractionResult));
                     ret.AddRange(GetTokenForRegexMatching(beforeString, config.MonthOfRegex, extractionResult));
+
+                    // Cases like "2 days from/before today/tomorrow"
+                    var isAgo = this.config.AgoRegex.Match(extractionResult.Text).Success;
```
Comment (guom08) [inline .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDatePeriodExtractor.cs]: Use AgoRegex (containing "before today") and LaterRegex (containing "from today") to judge whether it's cases like "2 days from/before today/tomorrow/yesterday").
If it falls into this category, judge whether it contains "more than" / "less than" / "within" prefix and added into the extract result

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDatePeriodExtractor.cs [hunk 6]
```diff
@@ -189,6 +190,32 @@ private List<Token> SingleTimePointWithPatterns(string text, DateObject referenc
...
+                        // cases like "within 3 days from yesterday/tomorrow" does not make any sense
+                        if (extractionResult.Text.Contains("today") || extractionResult.Text.Contains("now"))
+                        {
+                            var match = this.config.WithinNextPrefixRegex.Match(beforeString);
+                            if (match.Success)
+                            {
+                                var isNext = !string.IsNullOrEmpty(match.Groups["next"].Value);
+
+                                // cases like "within the next 5 days before today" is not acceptable
+                                if (!(isNext && isAgo))
```
Comment (guom08) [inline .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDatePeriodExtractor.cs]: "Next" and "Ago" should not occur at the same time.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseMergedDateTimeParser.cs [hunk 7]
```diff
@@ -738,28 +738,50 @@ public void AddAltPeriodToResolution(Dictionary<string, string> resolutionDic, s
...
-                // For the 'before' mod, the start of the period should be the end the new period, not the start 
+
+
+                // For the 'before' mod
+                // 1. Cases like "Before December", the start of the period should be the end of the new period, not the start
+                // 2. Cases like "More than 3 days before today", the date point should be the end of the new period
                 if (mod.Equals(Constants.BEFORE_MOD))
                 {
-                    res.Add(DateTimeResolutionKey.END, start);
+                    if (!string.IsNullOrEmpty(start) && !string.IsNullOrEmpty(end))
```
Comment (guom08) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseMergedDateTimeParser.cs]: Cases like "Before December", "Before the next two weeks". These cases contains a duration so it has "start" and "end".

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseMergedDateTimeParser.cs [hunk 8]
```diff
@@ -738,28 +738,50 @@ public void AddAltPeriodToResolution(Dictionary<string, string> resolutionDic, s
...
+                    {
+                        res.Add(DateTimeResolutionKey.END, start);
+                    }
+                    else
```
Comment (guom08) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseMergedDateTimeParser.cs]: Cases like "Within 2 days after today", "Less than 2 weeks after today".
"Within", "Less than" + datepoint. Datepoint don't have "start" and "end" at the same time before.

File: .NET/Microsoft.Recognizers.Text.DateTime/Utilities/AgoLaterUtil.cs [hunk 9]
```diff
@@ -22,8 +22,30 @@ public class AgoLaterUtil
...
 
-                if (MatchingUtil.GetTermIndex(beforeString, utilityConfiguration.InConnectorRegex, out index))
+                if (MatchingUtil.GetAgoLaterIndex(afterString, utilityConfiguration.AgoRegex, out index))
+                {
+                    // We don't support cases like "5 minutes from today" for now
+                    // Cases like "5 minutes ago" or "5 minutes from now" are supported
+                    // Cases like "2 days before today" or "2 weeks from today" are also supported
+                    var isDayMatchInAfterString = utilityConfiguration.AgoRegex.Match(afterString).Groups["day"].Success;
+
+                    if (!(isTimeDuration && isDayMatchInAfterString))
```
Comment (guom08) [inline .NET/Microsoft.Recognizers.Text.DateTime/Utilities/AgoLaterUtil.cs]: Filter out cases like "5 minutes from today"

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateExtractor.cs [hunk 10]
```diff
@@ -306,9 +306,16 @@ private List<Token> DurationWithBeforeAndAfter(string text, DateObject reference
...
                 // if it is a multiple duration and its type is not equal to Date than skip it.
-                if (er.Data != null && er.Data.ToString() != Constants.MultipleDuration_Date)
+                if (er.Data != null && er.Data.ToString().StartsWith("multipleDuration") && er.Data.ToString() != Constants.MultipleDuration_Date)
                 {
                     continue;
+                }
+
+                // special treatment for cases like "more than 3 days from today"
+                // should only extract "3 days from today" as datepoint
+                if (er.Data != null && (er.Data.ToString() == Constants.MORE_THAN_MOD || er.Data.ToString() == Constants.LESS_THAN_MOD))
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateExtractor.cs]: er = StripDurationInequality(er)
The if moves inside the renamed method.

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateExtractor.cs [hunk 11]
```diff
@@ -306,9 +306,16 @@ private List<Token> DurationWithBeforeAndAfter(string text, DateObject reference
...
             {
                 // if it is a multiple duration and its type is not equal to Date than skip it.
-                if (er.Data != null && er.Data.ToString() != Constants.MultipleDuration_Date)
+                if (er.Data != null && er.Data.ToString().StartsWith("multipleDuration") && er.Data.ToString() != Constants.MultipleDuration_Date)
                 {
                     continue;
+                }
+
+                // special treatment for cases like "more than 3 days from today"
+                // should only extract "3 days from today" as datepoint
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateExtractor.cs]: Change this comment to as you described offline.

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateExtractor.cs [hunk 12]
```diff
@@ -322,5 +329,30 @@ private List<Token> DurationWithBeforeAndAfter(string text, DateObject reference
             return ret;
         }
 
+        private void RemoveMoreThanOrLessThanPrefix(ExtractResult er)
+        {
+            var match = config.MoreThanRegex.Match(er.Text);
+            if (match.Success)
+            {
+                var originalLength = er.Text.Length;
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateExtractor.cs]: Nit: Simple refactor to reuse util code.

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDatePeriodExtractor.cs [hunk 13]
```diff
@@ -189,6 +190,32 @@ private List<Token> SingleTimePointWithPatterns(string text, DateObject referenc
...
+                    // Cases like "2 days from/before today/tomorrow"
+                    var isAgo = this.config.AgoRegex.Match(extractionResult.Text).Success;
+                    var isLater = this.config.LaterRegex.Match(extractionResult.Text).Success;
+
+                    if (isAgo || isLater)
+                    {
+                        ret.AddRange(GetTokenForRegexMatching(beforeString, config.LessThanRegex, extractionResult));
+                        ret.AddRange(GetTokenForRegexMatching(beforeString, config.MoreThanRegex, extractionResult));
+
+                        // cases like "within 3 days from yesterday/tomorrow" does not make any sense
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDatePeriodExtractor.cs]: Improve comment to describe the two cases for 'within'

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDatePeriodExtractor.cs [hunk 14]
```diff
@@ -189,6 +190,32 @@ private List<Token> SingleTimePointWithPatterns(string text, DateObject referenc
...
+                        if (extractionResult.Text.Contains("today") || extractionResult.Text.Contains("now"))
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDatePeriodExtractor.cs]: These words should get a special name and be in the YAML file.

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDatePeriodExtractor.cs [hunk 15]
```diff
@@ -189,6 +190,32 @@ private List<Token> SingleTimePointWithPatterns(string text, DateObject referenc
...
+                        {
+                            var match = this.config.WithinNextPrefixRegex.Match(beforeString);
+                            if (match.Success)
+                            {
+                                var isNext = !string.IsNullOrEmpty(match.Groups["next"].Value);
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDatePeriodExtractor.cs]: The group name should be in Constants.

Comment (guom08) [reply .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDatePeriodExtractor.cs]: Fixed.

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDatePeriodExtractor.cs (hunk 14 above)
Comment (guom08) [reply .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDatePeriodExtractor.cs]: Fixed.

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDatePeriodExtractor.cs (hunk 13 above)
Comment (guom08) [reply .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDatePeriodExtractor.cs]: Fixed.

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateExtractor.cs (hunk 12 above)
Comment (guom08) [reply .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateExtractor.cs]: Didn't find utility code doing similar thing.
Create a simple method to replace text using certain Regex and change corresponding files in ExtractResult, re-use this simple method to do the refactor.

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateExtractor.cs (hunk 11 above)
Comment (guom08) [reply .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateExtractor.cs]: Fixed

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateExtractor.cs (hunk 10 above)
Comment (guom08) [reply .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateExtractor.cs]: Fixed, here er is the iterative one, so create a "void" method doing similar work.

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateExtractor.cs [hunk 16]
```diff
@@ -306,9 +306,16 @@ private List<Token> DurationWithBeforeAndAfter(string text, DateObject reference
             foreach (var er in durationEr)
```
[the remark this answers is not in the document]
Comment (guom08) [reply .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateExtractor.cs]: Fixed

**Pull request merged**